XenModel/Actions/NRPE/NRPEUpdateAction.cs

```diff
@@ -0,0 +1,165 @@
+ * SUCH DAMAGE.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using XenAPI;
+
+
+namespace XenAdmin.Actions.NRPE
+{
+    public class NRPEUpdateAction : AsyncAction
+    {
+        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
+
+        private readonly NRPEHostConfiguration _nrpeOriginalConfig;  // NRPE configuration fetched from host
+        private readonly NRPEHostConfiguration _nrpeHostConfiguration;  // NRPE configuration after user modified
+
+        private readonly IXenObject _clone;
+
+        public NRPEUpdateAction(IXenObject host, NRPEHostConfiguration nrpeHostConfiguration, NRPEHostConfiguration nrpeOriginalConfig, bool suppressHistory)
+            : base(host.Connection, Messages.NRPE_ACTION_CHANGING, Messages.NRPE_ACTION_CHANGING, suppressHistory)
+        {
+            _clone = host;
+            _nrpeHostConfiguration = nrpeHostConfiguration;
+            _nrpeOriginalConfig = nrpeOriginalConfig;
+        }
+
+        protected override void Run()
+        {
+            if (_clone is Host)
+            {
+                SetNRPEConfigureForHost(_clone);
+            }
+            else
+            {
+                SetNRPEConfigureForPool();
+            }
+        }
+
+        private void SetNRPEConfigureForHost(IXenObject o)
+        {
+            // Enable/Disable NRPE
+            if (!_nrpeHostConfiguration.EnableNRPE == _nrpeOriginalConfig.EnableNRPE)
+            {
+                SetNRPEStatus(o, _nrpeHostConfiguration.EnableNRPE);
+            }
+            if (!_nrpeHostConfiguration.EnableNRPE) // If disable, return
+            {
+                return;
+            }
+
+            // NRPE General Configuration
+            if (!_nrpeHostConfiguration.AllowHosts.Equals(_nrpeOriginalConfig.AllowHosts)
+                || !_nrpeHostConfiguration.Debug.Equals(_nrpeOriginalConfig.Debug)
+                || !_nrpeHostConfiguration.SslLogging.Equals(_nrpeOriginalConfig.SslLogging))
+            {
+                SetNRPEGeneralConfiguration(o, _nrpeHostConfiguration.AllowHosts, _nrpeHostConfiguration.Debug, _nrpeHostConfiguration.SslLogging);
+            }
+
+            // NRPE Check Threshold
+            foreach (KeyValuePair<string, NRPEHostConfiguration.Check> kvp in _nrpeHostConfiguration.CheckDict)
+            {
+                NRPEHostConfiguration.Check currentCheck = kvp.Value;
+                _nrpeOriginalConfig.GetNRPECheck(kvp.Key, out NRPEHostConfiguration.Check OriginalCheck);
+                if (currentCheck != null && OriginalCheck != null
+                    && (!currentCheck.WarningThreshold.Equals(OriginalCheck.WarningThreshold)
+                    || !currentCheck.CriticalThreshold.Equals(OriginalCheck.CriticalThreshold)))
+                {
+                    SetNRPEThreshold(o, currentCheck.Name, currentCheck.WarningThreshold, currentCheck.CriticalThreshold);

```

Comment:
This Host.call_plugin supports to set only one threshold at a time.
Host.call_plugin takes a parameter thresholdArgDict because checkName, warningThreshold, and criticalThreshold are 3 keys for this call.

Reply:
Ok. Maybe we can consider it as a future enhancement to the plugin.